server/src/main/java/org/elasticsearch/action/support/nodes/BaseNodesRequest.java

```diff
@@ -81,4 +73,15 @@ public final void writeTo(StreamOutput out) throws IOException {
         // to send it out over the wire. Use a dedicated transport request just for the bits you need.
         TransportAction.localOnly();
     }
+
+    /**
+     * @return the nodes to which this request should fan out.
+     */
+    DiscoveryNode[] resolveNodes(ClusterState clusterState) {
+        assert nodesIds == null || concreteNodes == null;
+        return Objects.requireNonNullElseGet(
+            concreteNodes,
+            () -> Arrays.stream(clusterState.nodes().resolveNodes(nodesIds)).map(clusterState.nodes()::get).toArray(DiscoveryNode[]::new)
```

Comment:
Is this method used often?
It is surprising that we need to resolve nodeIds (that are actually can be names/hosts/`_local`) to the real nodeIds and then to DiscoveryNodes. Sounds like we should have a variant that returns a collection of DiscoveryNodes instead?

Reply:
Eh not really, it's pretty much just this spot. Not ideal for sure but this isn't a hot path.